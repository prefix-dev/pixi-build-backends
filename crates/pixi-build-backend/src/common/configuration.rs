use std::collections::BTreeMap;

use miette::IntoDiagnostic;
use pixi_build_types::PlatformAndVirtualPackages;
use rattler_build::{
    NormalizedKey,
    metadata::{Directories, PlatformWithVirtualPackages},
    recipe::variable::Variable,
};
use rattler_conda_types::ChannelUrl;
use rattler_virtual_packages::VirtualPackageOverrides;
use url::Url;

/// Returns the [`BuildConfigurationParams`] that will be used to construct a BuildConfiguration
pub fn build_configuration(
    channels: Vec<Url>,
    build_platform: Option<PlatformAndVirtualPackages>,
    host_platform: Option<PlatformAndVirtualPackages>,
    variant: BTreeMap<NormalizedKey, Variable>,
    directories: Directories,
) -> miette::Result<BuildConfigurationParams> {
    let build_platform = build_platform.map(|p| PlatformWithVirtualPackages {
        platform: p.platform,
        virtual_packages: p.virtual_packages.unwrap_or_default(),
    });

    let host_platform = host_platform.map(|p| PlatformWithVirtualPackages {
        platform: p.platform,
        virtual_packages: p.virtual_packages.unwrap_or_default(),
    });

    let (build_platform, host_platform) = match (build_platform, host_platform) {
        (Some(build_platform), Some(host_platform)) => (build_platform, host_platform),
        (build_platform, host_platform) => {
            let current_platform = rattler_build::metadata::PlatformWithVirtualPackages::detect(
                &VirtualPackageOverrides::from_env(),
            )
            .into_diagnostic()?;
            (
                build_platform.unwrap_or_else(|| current_platform.clone()),
                host_platform.unwrap_or(current_platform),
            )
        }
    };

    let channels = channels.into_iter().map(Into::into).collect();

    let params = BuildConfigurationParams {
        channels,
        build_platform,
        host_platform,
        variant,
        directories,
    };

    Ok(params)
}

/// The parameters used to construct a BuildConfiguration
#[derive(Debug)]
pub struct BuildConfigurationParams {
    pub channels: Vec<ChannelUrl>,
    pub build_platform: PlatformWithVirtualPackages,
    pub host_platform: PlatformWithVirtualPackages,
    pub variant: BTreeMap<NormalizedKey, Variable>,
    pub directories: Directories,
}
