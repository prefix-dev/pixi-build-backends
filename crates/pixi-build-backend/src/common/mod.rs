//! Common utilities that are shared between the different build backends.
mod configuration;
mod requirements;
mod variants;

pub use configuration::{BuildConfigurationParams, build_configuration};
pub use requirements::{PackageRequirements, SourceRequirements, requirements};
pub use variants::compute_variants;
