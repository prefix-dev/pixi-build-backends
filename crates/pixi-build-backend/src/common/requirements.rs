use std::collections::{BTreeMap, HashMap};

use rattler_build::{
    NormalizedKey,
    recipe::{parser::Requirements, variable::Variable},
};
use serde::Serialize;

use crate::{
    PackageSpec, ProjectModel, Targets, dependencies::ExtractedDependencies, traits::Dependencies,
};

pub struct PackageRequirements<P: ProjectModel> {
    /// Requirements for rattler-build
    pub requirements: Requirements,

    /// The source requirements
    pub source: SourceRequirements<P>,
}

#[derive(Debug, Serialize)]
#[serde(bound(
    serialize = "<<P::Targets as Targets>::Spec as PackageSpec>::SourceSpec: Serialize"
))]
pub struct SourceRequirements<P: ProjectModel> {
    /// Source package specification for build dependencies
    pub build: HashMap<String, <<P::Targets as Targets>::Spec as PackageSpec>::SourceSpec>,

    /// Source package specification for host dependencies
    pub host: HashMap<String, <<P::Targets as Targets>::Spec as PackageSpec>::SourceSpec>,

    /// Source package specification for runtime dependencies
    pub run: HashMap<String, <<P::Targets as Targets>::Spec as PackageSpec>::SourceSpec>,
}

/// Return requirements for the given project model
pub fn requirements<P: ProjectModel>(
    dependencies: Dependencies<<P::Targets as Targets>::Spec>,
    variant: &BTreeMap<NormalizedKey, Variable>,
) -> miette::Result<PackageRequirements<P>> {
    let build = ExtractedDependencies::from_dependencies(dependencies.build, variant)?;
    let host = ExtractedDependencies::from_dependencies(dependencies.host, variant)?;
    let run = ExtractedDependencies::from_dependencies(dependencies.run, variant)?;

    Ok(PackageRequirements {
        requirements: Requirements {
            build: build.dependencies,
            host: host.dependencies,
            run: run.dependencies,
            ..Default::default()
        },
        source: SourceRequirements {
            build: build.sources,
            host: host.sources,
            run: run.sources,
        },
    })
}
