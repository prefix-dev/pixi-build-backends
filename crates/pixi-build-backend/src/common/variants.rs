//! Variants trait
//!
use std::collections::BTreeMap;

use miette::IntoDiagnostic;
use rattler_build::{NormalizedKey, recipe::variable::Variable, variant_config::VariantConfig};
use rattler_conda_types::Platform;

use crate::ProjectModel;

/// Return variants for the given project model
pub fn compute_variants<P: ProjectModel>(
    project_model: &P,
    input_variant_configuration: Option<BTreeMap<NormalizedKey, Vec<Variable>>>,
    host_platform: Platform,
) -> miette::Result<Vec<BTreeMap<NormalizedKey, Variable>>> {
    // Create a variant config from the variant configuration in the parameters.
    let variant_config = VariantConfig {
        variants: input_variant_configuration.unwrap_or_default(),
        pin_run_as_build: None,
        zip_keys: None,
    };

    // Determine the variant keys that are used in the recipe.
    let used_variants = project_model.used_variants(Some(host_platform));

    // Determine the combinations of the used variants.
    variant_config
        .combinations(&used_variants, None)
        .into_diagnostic()
}
