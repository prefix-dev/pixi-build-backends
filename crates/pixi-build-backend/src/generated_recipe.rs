use std::{
    collections::{BTreeMap, BTreeSet},
    fmt::Debug,
    path::{Path, PathBuf},
};

use pixi_build_types::ProjectModelV1;
use rattler_build::{NormalizedKey, recipe::variable::Variable};
use rattler_conda_types::Platform;
use recipe_stage0::recipe::{IntermediateRecipe, Package, Value};
use serde::de::DeserializeOwned;

use crate::specs_conversion::from_targets_v1_to_conditional_requirements;

#[derive(Debug, Clone, Default)]
pub struct PythonParams {
    // Returns whetever the build is editable or not.
    // Default to false
    pub editable: bool,
}

/// The trait is responsible of converting a certain [`ProjectModelV1`] (or
/// others in the future) into an [`IntermediateRecipe`].
/// By implementing this trait, you can create a new backend for `pixi-build`.
///
/// It also uses a [`BackendConfig`] to provide additional configuration
/// options.
///
///
/// An instance of this trait is used by the [`IntermediateBackend`]
/// in order to generate the recipe.
pub trait GenerateRecipe {
    type Config: BackendConfig;

    /// Generates an [`IntermediateRecipe`] from a [`ProjectModelV1`].
    fn generate_recipe(
        &self,
        model: &ProjectModelV1,
        config: &Self::Config,
        manifest_path: PathBuf,
        // The host_platform will be removed in the future.
        // Right now it is used to determine if certain dependencies are present
        // for the host platform.
        // Instead, we should rely on recipe selectors and offload all the
        // evaluation logic to the rattler-build.
        host_platform: Platform,
        // Note: It is used only by python backend right now and may
        // be removed when profiles will be implemented.
        python_params: Option<PythonParams>,
    ) -> miette::Result<GeneratedRecipe>;

    /// Returns a list of globs that should be used to find the input files
    /// for the build process.
    /// For example, this could be a list of source files or configuration files
    /// used by Cmake.
    fn extract_input_globs_from_build(
        _config: &Self::Config,
        _workdir: impl AsRef<Path>,
        _editable: bool,
    ) -> BTreeSet<String> {
        BTreeSet::new()
    }

    /// Returns "default" variants for the given host platform. This allows
    /// backends to set some default variant configuration that can be
    /// completely overwritten by the user.
    ///
    /// This can be useful to change the default behavior of rattler-build with
    /// regard to compilers. But it also allows setting up default build
    /// matrices.
    fn default_variants(&self, _host_platform: Platform) -> BTreeMap<NormalizedKey, Vec<Variable>> {
        BTreeMap::new()
    }
}

pub trait BackendConfig: DeserializeOwned + Clone {
    /// At least debug dir should be provided by the backend config
    fn debug_dir(&self) -> Option<&Path>;

    /// Merge this configuration with a target-specific configuration.
    /// Target-specific values typically override base values.
    fn merge_with_target_config(&self, target_config: &Self) -> miette::Result<Self>;
}

#[derive(Default, Clone)]
pub struct GeneratedRecipe {
    pub recipe: IntermediateRecipe,
    pub metadata_input_globs: BTreeSet<String>,
    pub build_input_globs: BTreeSet<String>,
}

impl GeneratedRecipe {
    /// Creates a new [`GeneratedRecipe`] from a [`ProjectModelV1`].
    /// A default implementation that doesn't take into account the
    /// build scripts or other fields.
    pub fn from_model(model: ProjectModelV1) -> Self {
        let package = Package {
            name: Value::Concrete(model.name),
            version: Value::Concrete(
                model.version
                  .expect("`version` is required at the moment. In the future we will read this from `Cargo.toml`.")
                  .to_string(),
            ),
        };

        let requirements =
            from_targets_v1_to_conditional_requirements(&model.targets.unwrap_or_default());

        let ir = IntermediateRecipe {
            package,
            requirements,
            ..Default::default()
        };

        GeneratedRecipe {
            recipe: ir,
            ..Default::default()
        }
    }
}
