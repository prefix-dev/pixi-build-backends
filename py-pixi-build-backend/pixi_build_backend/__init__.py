"""
Python implementation of pixi-build-python backend using Python bindings.
"""
