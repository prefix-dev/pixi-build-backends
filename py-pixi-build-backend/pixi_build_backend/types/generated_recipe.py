from typing import Any, Dict, Optional, Protocol, List
from pathlib import Path
from pixi_build_backend.types.intermediate_recipe import IntermediateRecipe
from pixi_build_backend.pixi_build_backend import PyGeneratedRecipe, PyGenerateRecipe
from pixi_build_backend.types.platform import Platform
from pixi_build_backend.types.project_model import ProjectModelV1
from pixi_build_backend.types.python_params import PythonParams


class GeneratedRecipe:
    """A generated recipe wrapper."""

    _inner: PyGeneratedRecipe

    def __init__(self) -> None:
        self._inner = PyGeneratedRecipe()

    @classmethod
    def from_model(cls, model: ProjectModelV1, manifest_root: Path) -> "GeneratedRecipe":
        """Create a GeneratedRecipe from a ProjectModelV1."""
        instance = cls()
        instance._inner = PyGeneratedRecipe().from_model(model._inner, manifest_root)
        return instance

    @property
    def recipe(self) -> IntermediateRecipe:
        """Get the recipe."""
        return IntermediateRecipe._from_inner(self._inner.recipe)

    @property
    def metadata_input_globs(self) -> List[str]:
        """Get the metadata input globs."""
        return self._inner.metadata_input_globs

    @property
    def build_input_globs(self) -> List[str]:
        """Get the build input globs."""
        return self._inner.build_input_globs

    def __repr__(self) -> str:
        return self._inner.__repr__()

    def _into_py(self) -> PyGeneratedRecipe:
        """
        Converts this object into a type that can be used by the Rust code.
        """
        return self._inner


class GenerateRecipeProtocol(Protocol):
    """
    Protocol for generating recipes.
    This should be implemented by the Python generator.
    """

    def generate_recipe(
        self,
        model: ProjectModelV1,
        config: Dict[str, Any],
        manifest_path: str,
        host_platform: Platform,
        python_params: Optional[PythonParams] = None,
    ) -> GeneratedRecipe:
        """Generate a recipe for a Python package."""
        ...


class GenerateRecipe:
    """Protocol for generating recipes."""

    _inner: PyGenerateRecipe

    def __init__(self, instance: GenerateRecipeProtocol):
        self._inner = PyGenerateRecipe(instance)

    def _into_py(self) -> PyGenerateRecipe:
        """
        Converts this object into a type that can be used by the Rust code.
        """
        return self._inner
