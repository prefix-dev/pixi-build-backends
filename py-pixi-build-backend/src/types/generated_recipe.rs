use std::collections::BTreeSet;

use miette::IntoDiagnostic;
use pixi_build_backend::generated_recipe::{GenerateRecipe, GeneratedRecipe};
use pyo3::{
    Py, PyObject, PyResult, Python, pyclass, pymethods,
    types::{PyAnyMethods, PyString},
};
use recipe_stage0::recipe::IntermediateRecipe;

use crate::{
    create_py_wrap,
    recipe_stage0::recipe::PyIntermediateRecipe,
    types::{PyBackendConfig, PyPlatform, PyProjectModelV1, PyPythonParams},
};

create_py_wrap!(PyVecString, Vec<String>, |v: &Vec<String>,
                                           f: &mut std::fmt::Formatter<
    '_,
>| {
    write!(f, "[{}]", v.join(", "))
});

#[pyclass(get_all, set_all)]
#[derive(Clone)]
pub struct PyGeneratedRecipe {
    pub(crate) recipe: Py<PyIntermediateRecipe>,
    pub(crate) metadata_input_globs: Py<PyVecString>,
    pub(crate) build_input_globs: Py<PyVecString>,
}

#[pymethods]
impl PyGeneratedRecipe {
    #[new]
    pub fn new(py: Python) -> PyResult<Self> {
        Ok(PyGeneratedRecipe {
            recipe: Py::new(py, PyIntermediateRecipe::new(py)?)?,
            metadata_input_globs: Py::new(py, PyVecString::default())?,
            build_input_globs: Py::new(py, PyVecString::default())?,
        })
    }

    #[staticmethod]
    pub fn from_model(py: Python, model: PyProjectModelV1) -> Self {
        let recipe = GeneratedRecipe::from_model(model.inner.clone());
        let py_generated_recipe = PyIntermediateRecipe::from_intermediate_recipe(recipe.recipe, py);

        let metadata_input_globs_vec: Vec<String> =
            recipe.metadata_input_globs.into_iter().collect();
        let build_input_globs_vec: Vec<String> = recipe.build_input_globs.into_iter().collect();

        let metadata_input_globs_vec = PyVecString::from(metadata_input_globs_vec);
        let build_input_globs_vec = PyVecString::from(build_input_globs_vec);

        PyGeneratedRecipe {
            recipe: Py::new(py, py_generated_recipe).unwrap(),
            metadata_input_globs: Py::new(py, metadata_input_globs_vec).unwrap(),
            build_input_globs: Py::new(py, build_input_globs_vec).unwrap(),
        }
    }
}

impl PyGeneratedRecipe {
    pub fn to_generated_recipe(&self, py: Python) -> GeneratedRecipe {
        let recipe: IntermediateRecipe = self.recipe.borrow(py).to_intermediate_recipe(py);
        let metadata_input_globs: BTreeSet<String> =
            (*self.metadata_input_globs.borrow(py).clone())
                .clone()
                .into_iter()
                .collect();
        let build_input_globs: BTreeSet<String> = (*self.build_input_globs.borrow(py).clone())
            .clone()
            .into_iter()
            .collect();

        GeneratedRecipe {
            recipe,
            metadata_input_globs,
            build_input_globs,
        }
    }
}

/// Trait part
#[pyclass]
#[derive(Clone)]
pub struct PyGenerateRecipe {
    model: PyObject,
}

#[pymethods]
impl PyGenerateRecipe {
    #[new]
    pub fn new(model: PyObject) -> Self {
        PyGenerateRecipe { model }
    }
}

impl GenerateRecipe for PyGenerateRecipe {
    type Config = PyBackendConfig;

    fn generate_recipe(
        &self,
        model: &pixi_build_types::ProjectModelV1,
        config: &Self::Config,
        manifest_path: std::path::PathBuf,
        host_platform: rattler_conda_types::Platform,
        python_params: Option<pixi_build_backend::generated_recipe::PythonParams>,
    ) -> miette::Result<pixi_build_backend::generated_recipe::GeneratedRecipe> {
        let recipe: GeneratedRecipe = Python::with_gil(|py| {
            let manifest_str = manifest_path.to_string_lossy().to_string();

            // we don't pass the wrapper but the python inner model directly
            let py_object = config.model.clone();

            // For other types, we try to wrap them into the Python class
            // So user can use the Python API
            let project_model_class = py
                .import("pixi_build_backend.types.project_model")
                .into_diagnostic()?
                .getattr("ProjectModelV1")
                .into_diagnostic()?;

            let project_model = project_model_class
                .call_method1("_from_py", (PyProjectModelV1::from(model),))
                .into_diagnostic()?;

            let platform_model_class = py
                .import("pixi_build_backend.types.platform")
                .into_diagnostic()?
                .getattr("Platform")
                .into_diagnostic()?;

            let platform_model = platform_model_class
                .call_method1("_from_py", (PyPlatform::from(host_platform),))
                .into_diagnostic()?;

            let python_params_class = py
                .import("pixi_build_backend.types.python_params")
                .into_diagnostic()?
                .getattr("PythonParams")
                .into_diagnostic()?;
            let python_params_model = python_params_class
                .call_method1(
                    "_from_py",
                    (PyPythonParams::from(python_params.unwrap_or_default()),),
                )
                .into_diagnostic()?;

            let generated_recipe_py = self
                .model
                .bind(py)
                .call_method(
                    "generate_recipe",
                    (
                        project_model,
                        py_object,
                        PyString::new(py, manifest_str.as_str()),
                        platform_model,
                        python_params_model,
                    ),
                    None,
                )
                .into_diagnostic()?;

            // To expose a nice API for the user, we extract the PyGeneratedRecipe
            // calling private _into_py method
            let generated_recipe: PyGeneratedRecipe = generated_recipe_py
                .call_method0("_into_py")
                .into_diagnostic()?
                .extract::<PyGeneratedRecipe>()
                .into_diagnostic()?;

            Ok::<_, miette::Report>(generated_recipe.to_generated_recipe(py))
        })?;

        Ok(recipe)
    }
}
