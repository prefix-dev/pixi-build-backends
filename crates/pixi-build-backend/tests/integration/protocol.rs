use crate::common::model::{convert_test_model_to_project_model_v1, load_project_model_from_json};
use imp::TestGenerateRecipe;
use pixi_build_backend::{intermediate_backend::IntermediateBackend, protocol::Protocol};
use pixi_build_types::{
    ChannelConfiguration, PlatformAndVirtualPackages,
    procedures::{conda_build::CondaBuildParams, conda_metadata::CondaMetadataParams},
};
use rattler_build::console_utils::LoggingOutputHandler;
use rattler_conda_types::Platform;
use serde_json::json;
use tempfile::TempDir;
use url::Url;

#[cfg(test)]
mod imp {
    use std::path::{Path, PathBuf};

    use pixi_build_backend::generated_recipe::{
        BackendConfig, GenerateRecipe, GeneratedRecipe, PythonParams,
    };
    use serde::Deserialize;

    #[derive(Debug, Default, Deserialize, Clone)]
    #[serde(rename_all = "kebab-case")]
    pub struct TestBackendConfig {
        /// If set, internal state will be logged as files in that directory
        pub debug_dir: Option<PathBuf>,
    }

    #[cfg(test)]
    #[derive(Clone, Default)]
    pub(crate) struct TestGenerateRecipe {}

    impl BackendConfig for TestBackendConfig {
        fn debug_dir(&self) -> Option<&Path> {
            self.debug_dir.as_deref()
        }
    }

    impl GenerateRecipe for TestGenerateRecipe {
        type Config = TestBackendConfig;

        fn generate_recipe(
            &self,
            model: &pixi_build_types::ProjectModelV1,
            _config: &Self::Config,
            manifest_path: PathBuf,
            _host_platform: rattler_conda_types::Platform,
            _python_params: Option<PythonParams>,
        ) -> miette::Result<GeneratedRecipe> {
            let generated_recipe = GeneratedRecipe::from_model(model.clone(), manifest_path);
            Ok(generated_recipe)
        }
    }
}

#[tokio::test]
async fn test_conda_get_metadata() {
    let tmp_dir = TempDir::new().unwrap();
    let tmp_dir_path = tmp_dir.path().to_path_buf();

    let pixi_manifest = tmp_dir_path.join("pixi.toml");

    let build_dir = tmp_dir_path.join("build");

    // Load a model from JSON
    let original_model = load_project_model_from_json("minimal_project_model.json");

    // Serialize it back to JSON
    let project_model_v1 = convert_test_model_to_project_model_v1(original_model);

    // save the pixi.toml file to a temporary location
    fs_err::write(&pixi_manifest, toml::to_string(&project_model_v1).unwrap()).unwrap();

    let platform = PlatformAndVirtualPackages {
        platform: Platform::Linux64,
        virtual_packages: None,
    };

    let channel_configuration = ChannelConfiguration {
        base_url: Url::parse("https://prefix.dev").unwrap(),
    };

    let channel_base_urls = vec![Url::parse("https://prefix.dev/conda-forge").unwrap()];

    let params = CondaMetadataParams {
        build_platform: Some(platform.clone()),
        host_platform: Some(platform.clone()),
        channel_base_urls: Some(channel_base_urls),
        channel_configuration,
        variant_configuration: None,
        work_directory: build_dir,
    };

    let some_config = json!({
        "debug-dir": "some_debug_dir",
    });

    let intermediate_backend = IntermediateBackend::<TestGenerateRecipe>::new(
        pixi_manifest.clone(),
        project_model_v1,
        TestGenerateRecipe::default(),
        some_config,
        LoggingOutputHandler::default(),
        None,
    )
    .unwrap();

    let conda_metadata = intermediate_backend
        .conda_get_metadata(params)
        .await
        .unwrap();

    insta::assert_yaml_snapshot!(conda_metadata, {
        ".packages[0].sources.boltons.path" => "[redacted]",
        ".packages[0].subdir" => "[redacted]",
    })
}

#[tokio::test]
async fn test_conda_build() {
    let tmp_dir = TempDir::new().unwrap();
    let tmp_dir_path = tmp_dir.path().to_path_buf();

    let pixi_manifest = tmp_dir_path.join("pixi.toml");
    let build_dir = tmp_dir_path.join("build");

    // Load a model from JSON
    let original_model = load_project_model_from_json("minimal_project_model_for_build.json");

    // Serialize it back to JSON
    let project_model_v1 = convert_test_model_to_project_model_v1(original_model);

    // save the pixi.toml file to a temporary location
    fs_err::write(&pixi_manifest, toml::to_string(&project_model_v1).unwrap()).unwrap();

    let channel_configuration = ChannelConfiguration {
        base_url: Url::parse("https://prefix.dev").unwrap(),
    };

    let channel_base_urls = vec![Url::parse("https://prefix.dev/conda-forge").unwrap()];

    let build_params = CondaBuildParams {
        build_platform_virtual_packages: None,
        host_platform: None,
        channel_base_urls: Some(channel_base_urls),
        channel_configuration,
        outputs: None,
        variant_configuration: None,
        work_directory: build_dir.clone(),
        editable: false,
    };

    let some_config = json!({
        "debug-dir": "some_debug_dir",
    });

    let intermediate_backend = IntermediateBackend::new(
        pixi_manifest.clone(),
        project_model_v1,
        TestGenerateRecipe::default(),
        some_config,
        LoggingOutputHandler::default(),
        None,
    )
    .unwrap();

    let conda_build_result = intermediate_backend
        .conda_build(build_params)
        .await
        .unwrap();

    insta::assert_yaml_snapshot!(conda_build_result, {
        ".packages[0].output_file" => "[redacted]",
        ".packages[0].subdir" => "[redacted]",
    });
}
