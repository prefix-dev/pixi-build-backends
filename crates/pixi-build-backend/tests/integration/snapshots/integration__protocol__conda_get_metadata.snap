---
source: crates/pixi-build-backend/tests/integration/protocol.rs
expression: conda_metadata
---
packages:
  - name: minimal-package
    version: 1.0.0
    build: hbf21a9e_0
    buildNumber: 0
    subdir: "[redacted]"
    depends:
      - python >=3.8
    constraints: []
    license: ~
    licenseFamily: ~
    noarch: false
    sources: {}
inputGlobs: []
